### PR TITLE
Add ExecutorService backed by SchedulerService

### PR DESCRIPTION
**Pull request by Kiskae**
This PR was previously commited as PR #829 and was merged into the old java8 branch, but it was not included in the final java8 update.
So this PR updates that PR so it can be added to the master branch.

SpongeCommon PR: https://github.com/SpongePowered/SpongeCommon/pull/200

This PR will also satisfy the requirements for #840

### CODE REVIEW

File: src/main/java/org/spongepowered/api/service/scheduler/SchedulerService.java [hunk 1]
```diff
@@ -81,4 +81,26 @@
      * @return A set of scheduled tasks
      */
     Set<Task> getScheduledTasks(Object plugin);
+
+
```
Comment (Phroa) [inline src/main/java/org/spongepowered/api/service/scheduler/SchedulerService.java]: Double blank line.

File: src/main/java/org/spongepowered/api/service/scheduler/SpongeExecutorService.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+import java.util.concurrent.Callable;
+import java.util.concurrent.RunnableScheduledFuture;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A delegating ExecutionService that schedules all its tasks on
+ * Sponge's {@link SchedulerService}.
+ * <p>
```
Comment (Phroa) [inline src/main/java/org/spongepowered/api/service/scheduler/SpongeExecutorService.java]: Style is a blank line before a `<p>` and no indentation for their contents (e.g. `<p> content and tags are inline </p>`)

File: src/main/java/org/spongepowered/api/service/scheduler/SpongeExecutorService.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+    @Override
+    SpongeFuture<?> scheduleAtFixedRate(Runnable command, long initialDelay, long period, TimeUnit unit);
+
+    @Override
+    SpongeFuture<?> scheduleWithFixedDelay(Runnable command, long initialDelay, long delay, TimeUnit unit);
+
+    interface SpongeFuture<V> extends RunnableScheduledFuture<V> {
+
+        /**
```
Comment (Phroa) [inline src/main/java/org/spongepowered/api/service/scheduler/SpongeExecutorService.java]: This method needs a summary sentence

**Comment by Kiskae** [comment]
@Phroa handled, thanks for the comments

**Comment by Phroa** [comment]
:heart:

File: src/main/java/org/spongepowered/api/service/scheduler/SpongeExecutorService.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+         * Retrieves the {@link SchedulerService} {@link Task} instance that
+         * is responsible for the execution of this future.
+         *
+         * @return The backing task
+         */
+        Task getTask();
+    }
+}
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/service/scheduler/SpongeExecutorService.java]: Missing newline

Comment (Kiskae) [inline src/main/java/org/spongepowered/api/service/scheduler/SpongeExecutorService.java]: @simon816 fixed

**Comment by Kiskae** [comment]
API and implementation have been rebased onto `SchedulerService` changes

**Comment by gabizou** [comment]
@Kiskae can you fix the errors causing Travis to fail?

**Comment by Kiskae** [comment]
@gabizou Rebased and errors have been fixed

**Comment by Kiskae** [comment]
I've included the binding annotations written by @Pieguy128 from PR https://github.com/SpongePowered/SpongeAPI/pull/840 since it heavily depends on this PR and it only required minimal changes to the PR.

File: src/main/java/org/spongepowered/api/service/scheduler/SpongeExecutorService.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    @Override
+    <V> SpongeFuture<V> schedule(Callable<V> callable, long delay, TimeUnit unit);
+
+    @Override
+    SpongeFuture<?> scheduleAtFixedRate(Runnable command, long initialDelay, long period, TimeUnit unit);
+
+    @Override
+    SpongeFuture<?> scheduleWithFixedDelay(Runnable command, long initialDelay, long delay, TimeUnit unit);
+
+    interface SpongeFuture<V> extends RunnableScheduledFuture<V> {
```
Comment (Xfel) [inline src/main/java/org/spongepowered/api/service/scheduler/SpongeExecutorService.java]: Is there a specific reason for this class extending RunnableScheduledFuture? I don't see the use case where I would actually want to call the run method on that task, so I'm wondering why you would expose this.

Comment (Kiskae) [inline src/main/java/org/spongepowered/api/service/scheduler/SpongeExecutorService.java]: Mostly to keep the interface as flexible as possible, if the users wants to execute the future before the scheduler can do it then they should be able to.

It also mirrors some of the access shown with the `Task` class, which allows access to the internal `Consumer` and can be executed much like `run()` does in `RunnableFuture`

**Comment by gratimax** [comment]
This looks fantastic, but has been sitting unchanged for a while. @SpongePowered/developers any more concerns?

**Comment by Kiskae** [comment]
@gratimax I actually had one last idea that could be added based on https://github.com/SpongePowered/SpongeAPI/issues/892
Basically document that the object passed to `create(A)syncExecutor` can be either a plugin object or plugin container and adjusting accordingly in the implementation

**Comment by Kiskae** [comment]
Although that probably should be done as a seperate consistent change throughout the project where plugin objects are used to avoid creating special cases

**Comment by zml2008** [comment]
@Kiskae That's already supported -- it's implemented in the plugin manager.

**Comment by Xfel** [comment]
However, using this feature, you could omit the getInstance() calls on the PluginContainers and pass them directly - which would save us two map lookups.

**Comment by Kiskae** [comment]
@zml2008 But its not specified in the javadocs: https://github.com/SpongePowered/SpongeAPI/blob/20bad1be5c4b80598d820b4c937c6390d0b0384b/src/main/java/org/spongepowered/api/plugin/PluginManager.java#L40 If that method can accept PluginContainers as well then it should probably be documented so implementers can rely on it (Since the javadocs only specify 'instance', which refers to '{@link Plugin} instances' according to the javadocs).

File: src/main/java/org/spongepowered/api/service/scheduler/AsynchronousExecutor.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+ * Provides a convenient injection for {@link SchedulerService#createAsyncExecutor(Object)}.
+ *
+ * <p>Use this annotation on a {@link SpongeExecutorService}.
+ * Remember that {@link Inject} is also necessary.</p>
+ */
+@BindingAnnotation
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.PARAMETER, ElementType.FIELD})
+public @interface AsynchronousExecutor {
+}
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/service/scheduler/AsynchronousExecutor.java]: Missing newline

File: src/main/java/org/spongepowered/api/service/scheduler/SynchronousExecutor.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+ * Provides a convenient injection for {@link SchedulerService#createSyncExecutor(Object)}.
+ *
+ * <p>Use this annotation on a {@link SpongeExecutorService}.
+ * Remember that {@link Inject} is also necessary.</p>
+ */
+@BindingAnnotation
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.PARAMETER, ElementType.FIELD})
+public @interface SynchronousExecutor {
+}
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/service/scheduler/SynchronousExecutor.java]: Missing newline

File: src/main/java/org/spongepowered/api/service/scheduler/AsynchronousExecutor.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+import javax.inject.Inject;
+
+/**
+ * Provides a convenient injection for {@link SchedulerService#createAsyncExecutor(Object)}.
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/service/scheduler/AsynchronousExecutor.java]: Maximum line length for javadocs is 80 characters.

File: src/main/java/org/spongepowered/api/service/scheduler/SchedulerService.java [hunk 9]
```diff
@@ -89,4 +89,25 @@
      * @return The preferred tick interval
      */
     int getPreferredTickInterval();
+
+    /**
+     * Creates a new ExecutorService that can be used to schedule synchronous
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/service/scheduler/SchedulerService.java]: use `{@link ExecutorService}`

File: src/main/java/org/spongepowered/api/service/scheduler/SchedulerService.java [hunk 10]
```diff
@@ -89,4 +89,25 @@
...
+     * tasks through the standard Java concurrency interfaces.
+     *
+     * @param plugin The plugin that will own the created tasks
+     * @return A new executor service that can be used to execute
+     *         synchronous tasks
+     */
+    SpongeExecutorService createSyncExecutor(Object plugin);
+
+    /**
+     * Creates a new ExecutorService that can be used to schedule asynchronous
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/service/scheduler/SchedulerService.java]: use `{@link ExecutorService}`

File: src/main/java/org/spongepowered/api/service/scheduler/SpongeExecutorService.java [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+ */
+package org.spongepowered.api.service.scheduler;
+
+import java.util.concurrent.Callable;
+import java.util.concurrent.RunnableScheduledFuture;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A delegating ExecutionService that schedules all its tasks on
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/service/scheduler/SpongeExecutorService.java]: ExecutionService? Probably mean `{@link ExecutorService}`

File: src/main/java/org/spongepowered/api/service/scheduler/SynchronousExecutor.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+import javax.inject.Inject;
+
+/**
+ * Provides a convenient injection for {@link SchedulerService#createSyncExecutor(Object)}.
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/service/scheduler/SynchronousExecutor.java]: Maximum line length for javadocs is 80 characters.

**Comment by Kiskae** [comment]
@simon816 Javadocs have been adjusted :+1:

**Comment by simon816** [comment]
@SpongePowered/developers can this be given the go-ahead? Seems good to me

**Comment by gabizou** [comment]
Correct me if I'm wrong but, this service seems to be reliant that it's not going to be replaced by plugins, is this correct?

If so, please name it something aside from service as core services will be renamed very soon to note their difference from replaceable services.

**Comment by simon816** [comment]
@gabizou I assume it's following the naming scheme from `java.util.concurrent.ScheduledExecutorService`

**Comment by Kiskae** [comment]
@gabizou as @simon816 says, its not a Service in the sense that it is a part of Sponge' ServiceManager. I'm following the naming scheme of the java API regarding the various executors.

**Pull request closed without merging**